src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+module Generalize (generalizeArgs, generalizeIxDict, Generalized) where
+
+import Control.Monad
+import Control.Category ((>>>))
+
+import Syntax
+import Inference
+import QueryType
+import Name
+import Err
+import MTL1
+import LabeledItems
+
+type Generalized (e::E) (n::S) = (Abs (Nest Binder) e n, [Atom n])
+
+generalizeIxDict :: EnvReader m => Dict n -> m n (Generalized Dict n)
+generalizeIxDict dict = liftGeneralizerM do
+  dict' <- sinkM dict
+  dictTy <- getType dict'
+  dictTyGeneralized <- generalizeType dictTy
+  liftEnvReaderM $ generalizeDict dictTyGeneralized dict'
+{-# INLINE generalizeIxDict #-}
+
+generalizeArgs ::EnvReader m => Nest PiBinder n l -> [Atom n] -> m n (Generalized (ListE Atom) n)
+generalizeArgs reqTys allArgs =  liftGeneralizerM $ runSubstReaderT idSubst do
+  PairE (Abs reqTys' UnitE) (ListE allArgs') <- sinkM $ PairE (Abs reqTys UnitE) (ListE allArgs)
+  ListE <$> go reqTys' allArgs'
+  where
+    go ::Nest PiBinder i i' -> [Atom n] -> SubstReaderT AtomSubstVal GeneralizerM i n [Atom n]
+    go Empty [] = return []
+    go (Nest (PiBinder b ty _) bs) (arg:args) = do
+      ty' <- substM ty
+      arg' <- case ty' of
+        TyKind   -> liftSubstReaderT $ generalizeType arg
+        DictTy _ -> generalizeDict ty' arg
+        _ -> return arg
+      args'' <- extendSubst (b@>SubstVal arg') $ go bs args
+      return $ arg' : args''
+    go _ _ = error "zip error"
+{-# INLINE generalizeArgs #-}
+
+-- === generalizer monad plumbing ===
+
+data GeneralizationEmission n l = GeneralizationEmission (Binder n l) (Atom n)
+type GeneralizationEmissions = RNest GeneralizationEmission
+
+newtype GeneralizerM n a = GeneralizerM {
+  runGeneralizerM' :: DoubleInplaceT Env GeneralizationEmissions UnitB HardFailM n a }
```

Comment:
Why use `DoubleInplaceT` when one of the emissions are `UnitB`? Isn't that equivalent to the regular `InplaceT`?

Reply:
Actually no! DoubleInplaceT has two sets of emissions, but they behave differently. The `d1` emissions are emitted to the "top" (really, the level at which `runDoubleINplaceT` was called), and the emission can fail because it requires hoisting. Ideally, we'd have `InplaceT` and `HoistingInplaceT` and somehow `DoubleInplaceT` would be the composition of those two things. But I haven't figured out how to make that work.

Reply:
Sure, I get that, but if the "top" emissions are `UnitB` then they always successfully hoist and don't give you any extra information at the top. That's why I thought it ends up being equivalent to `InplaceT`? If not, what am I missing?

Reply:
Ah, I see the confusion. Here, we're using `UnitB` for the "bottom" emissions. The top emissions are `GeneralizationEmissions`.